KAFKA-12497: Skip periodic offset commits for failed source tasks

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-12497)

This change serves two purposes:
1. Eliminate unnecessary log messages for offset commit of tasks that don't need to perform offset commits (e.g., a task that has failed and for which all data has been flushed and committed)
2. Stop blocking the offset commit thread unnecessarily for flushes that will never succeed because the task's producer has failed to send a record in the current batch with a non-retriable error

Existing unit tests for the `OffsetStorageWriter` are tweaked to verify the small change made to it. Several new unit tests are added for the `WorkerSourceTask` that cover various cases where offset commits should not be attempted, and some existing tests are modified to cover cases where offset commits either should or should not be attempted.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-12497: Skip unnecessary source task offset commits

C0urante commented:
@ncliang @gharris1727 @kpatelatwork @ddasarathan could one or two of you take a look at this when you have time?

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -370,8 +372,9 @@ private boolean sendRecords() {
                         if (e != null) {
                             log.error("{} failed to send record to {}: ", WorkerSourceTask.this, topic, e);
-                            log.debug("{} Failed record: {}", WorkerSourceTask.this, preTransformRecord);
+                            log.trace("{} Failed record: {}", WorkerSourceTask.this, preTransformRecord);
```
Not required for this PR but important to downgrade this to `TRACE` since it contains record contents and we do not log record contents above that level.

kpatelatwork commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:
```diff
@@ -105,6 +105,11 @@ public void remove(ConnectorTaskId id) {
     }
 
     private void commit(WorkerSourceTask workerTask) {
+        if (!workerTask.shouldCommitOffsets()) {
```
@C0urante  overall the PR looks good but somehow I feel this `shouldCommitOffsets` is exposing internal details and may be `commitOffsets` should internally encapsulate this.  

How about we move `shouldCommitOffsets` check as first line in the `commitOffsets` method. and remove the `log.debug("{} Committing offsets", workerTask); `from this method as its redundant and `workerTask .commitOffsets` is already doing it.  Your only challenge would be `return false` is in a lot of places in `workerTask .commitOffsets` but we can may be convert the current method to a private method and add a new commitOffsets wrapper method to make that easy. We will also need to move `log.error("{} Failed to commit offsets", workerTask);` in the wrapper method.  This way all we need to do is call `workerTask.commitOffsets()` in try/catch here.

C0urante replied:
I initially wanted to do just that, but held off for two reasons:

1. The `SourceTaskOffsetCommitter` log message may be valuable to some users, especially if they've muted the `WorkerSourceTask` namespace or at least raised its level to `INFO`.
2. It's a little easier to test this logic separately by creating a separate method that can be queried on its own. The `boolean` return value of `WorkerSourceTask::commitOffsets` makes it difficult to determine if an offset commit has been attempted and failed, or skipped entirely.

I'm not super attached to either approach; if you have thoughts on why one or the other is better I'm happy to hear them.

kpatelatwork replied:
I prefer the encapsulation approach but will leave the decision to you as it's a minor nitpick :).

I prefer encapsulation because the caller to `commitOffsets` need not worry about first calling `shouldCommitOffsets` and only if it's true then I should call `commitOffsets`.  Think like we don't need to do  `!ArrayList.isEmpty` check before iterating over it :).

ncliang replied:
I tend to agree with @kpatelatwork that the `shouldCommit` method is better encapsulated in `WorkerSourceTask` . There is a call site within `WorkerSourceTask` that could also benefit from the `shouldCommit` check, for instance. https://github.com/apache/kafka/blob/a63e5be4195e97e5b825b5912291144d2d0283a3/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L266

For the testing aspect, could you make the method visible to tests to do the assertion?

C0urante replied:
So a few things here:

1. Even if `commitOffsets` is called without the `shouldCommitOffsets` check, if any of the conditions that lead to `shouldCommitOffsets` returning `false` arise in the middle of that call to `commitOffsets`, the offset commit will be prematurely aborted and, at that point, effectively become a no-op. So with the analogy of iterating over a list--this is actually perfectly in line with that example; you don't _have_ to call `List::isEmpty` before iterating since iterating over an empty list is effectively a no-op, but you can choose to if you'd like.
2. We could add a `shouldCommitOffsets` guard against the end-of-life offset commit, but since that only occurs once per task and, as mentioned above, it'll now effectively be a no-op if offset commit isn't possible, I don't think it's really necessary. The points of this PR are to avoid cluttering up the single-threaded source task offset committer and to avoid continually emitting useless and confusing log messages. Gating on a one-time offset commit that only takes place once the task is shut down isn't really necessary for either of these.
3. The `SourceTaskOffsetCommitter::commit` method's "{} Committing offsets" log message is super old and I'm really hesitant to mess with that, even by moving it into a separate class. I know this seems trivial to developers but logs are often the first or second line of defense users have in diagnosing incorrect or unexpected behavior on their workers and I don't want to make that any harder, especially for users with log processing pipelines that parse logging namespaces from messages and allow users to search based on that information.

kpatelatwork replied:
makes sense

kpatelatwork commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java:
```diff
@@ -120,6 +120,13 @@ public synchronized boolean beginFlush() {
+    /**
+     * @return whether there's anything to flush right now.
+     */
+    public boolean willFlush() {
```
should this method be synchronized?  All  other access to `data` seems to be in synchronized methods

C0urante replied:
Good catch, thanks!

kpatelatwork commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java:
```diff
@@ -662,20 +666,94 @@ public void testSendRecordsProducerCallbackFail() throws Exception {
+            5. The first offset commit succeeds as the first record has been sent successfully
+            6. No further offset commits are attempted as the only remaining record has failed with a non-retriable error
+         */
+        createWorkerTask();
```
two calls to createWorkerTask , is this a  copy paste or I am missing something :)?

C0urante replied:
Whoops! Thanks for catching this.

kpatelatwork commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java:
```diff
@@ -662,20 +666,94 @@ public void testSendRecordsProducerCallbackFail() throws Exception {
+            5. The first offset commit fails even though first record has been sent successfully, (possibly from failure to produce offsets to Kafka)
+            6. No further offset commits are attempted as the new record batch contains the second record, which has failed with a non-retriable error
+         */
+        createWorkerTask();
```
two calls  to createWorkerTask

C0urante replied:
Whoops! Thanks for catching this.

kpatelatwork commented:
@C0urante  overall looks good but as I am new to the code, I tried my best to review but I recommend getting a LGTM from one more reviewer also who knows this code better than me to ensure we don't miss something obvious?

ncliang commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -482,6 +486,22 @@ private synchronized void recordSent(final ProducerRecord<byte[], byte[]> record
+            currentBatchFailed = true;
+            if (flushing) {
+                // flush thread may be waiting on the outstanding messages to clear
+                this.notifyAll();
```
This implementation differs from the implementation of `recordSend` above when `outstandingMessages` contains records other than the current record. `recordSend` will wait until `outstandingMessages` is empty and `flushing` before `notifyAll` . Is this intentional because we've failed and can't clear the outstanding messages?

C0urante replied:
> Is this intentional because we've failed and can't clear the outstanding messages?

Yep, exactly. We only need to `notifyAll` here in order to potentially wake up an offset commit thread that's parked [here](https://github.com/C0urante/kafka/blob/01797a5004838d970db21a54bd99a29c7c63f5d4/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L538) waiting for all outstanding messages to be flushed. If the failed message isn't a part of that batch, or we're not in the middle of flushing at all, no wakeup is necessary.

ncliang reviewed (commented):
Thanks @C0urante . I also agree with @kpatelatwork that it is better to encapsulate the should offset commit logic in `WorkerSourceTask` . Otherwise I think generally LGTM.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:
```diff
@@ -105,6 +105,11 @@ public void remove(ConnectorTaskId id) {
 
     private void commit(WorkerSourceTask workerTask) {
+        if (!workerTask.shouldCommitOffsets()) {
+            log.trace("{} Skipping offset commit as there is nothing to be committed", workerTask);
```
The method used in the `if` is called `shouldCommitOffsets`, not `areThereOffsetsToBeCommitted`, so is this message accurate, or perhaps the method name is slightly misleading? Maybe amending the message to "...there are no offsets that should be committed"?

C0urante replied:
Sure, the message rewording is fair. Renaming to `areThereOffsetsToBeCommitted` is a little verbose and although it does technically capture the case where the task's producer has failed to send a record with a non-retriable error, I think it may still be a little misleading.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -482,6 +486,22 @@ private synchronized void recordSent(final ProducerRecord<byte[], byte[]> record
+        }
+    }
+
+    public boolean shouldCommitOffsets() {
```
Can we add some Javadoc

C0urante replied:
Ack, added.

tombentley commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java:
```diff
@@ -214,6 +215,12 @@ public void testCancelAfterAwaitFlush() throws Exception {
+    private void assertBeginFlush(boolean shouldFlush) {
+        Consumer<Boolean> assertion = shouldFlush ? Assert::assertTrue : Assert::assertFalse;
+        assertion.accept(writer.willFlush());
+        assertion.accept(writer.beginFlush());
```
Wouldn't `assertEquals(shouldFlush, writer.willFlush())` and `assertEquals(shouldFlush, writer.beginFlush())` be clearer?

C0urante replied:
Yeah, good point. Thought it would make messages clearer for failed assertions but the difference isn't worth the unreadable code.

tombentley reviewed (commented):
@C0urante I've taken a very quick initial look. Given the JIRA is explicitly about logging it would be nice if there were assertions in the tests that directly checked for logging. Could we use a LogCaptureAppender to make such assertions?

tombentley commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java:
```diff
@@ -1771,4 +1772,20 @@ private void expectTopicCreation(String topic) {
+        LogCaptureAppender.setClassLoggerToTrace(SourceTaskOffsetCommitter.class);
+        LogCaptureAppender.setClassLoggerToTrace(WorkerSourceTask.class);
+        try (LogCaptureAppender committerAppender = LogCaptureAppender.createAndRegister(SourceTaskOffsetCommitter.class)) {
+            try (LogCaptureAppender taskAppender = LogCaptureAppender.createAndRegister(WorkerSourceTask.class)) {
```
Any reason not to use multi-resource try-with-resources?

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -506,9 +530,10 @@ public boolean commitOffsets() {
+                    // outstanding messages, so we can stop flushing immediately. Similarly, if the task has been cancelled, no more records will be
+                    // sent from the producer; in that case, if any outstanding messages remain, we can also stop flushing immediately
+                    if (currentBatchFailed || isCancelled() || timeoutMs <= 0) {
                         log.error("{} Failed to flush, timed out while waiting for producer to flush outstanding {} messages", this, outstandingMessages.size());
```
The "timed out" part of the error is a little misleading now.

tombentley reviewed (commented):
@C0urante the extra assertions in the test look reasonable to me.

C0urante commented:
I've rebased onto the latest trunk. I think with the latest changes (especially [these logging improvements](https://github.com/apache/kafka/blob/a3ab7d5b42d02f194a8bdf3047c374f938ca0dad/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L479-L504)) most of the changes in this PR were made redundant. The only remaining room for improvement IMO is skipping log messages for failed tasks; the other issues (squatting on the source task offset commit thread too long for failed messages to be acknowledged, and misleading users with messages about flushing 0 records) have already been addressed.

I've force-pushed a single commit that brings this PR up to date with the latest trunk; going to push an additional commit later this week that addresses the review comments that have been left on it.

C0urante committed:
Move LogCaptureAppender to common package in clients module

C0urante committed:
Replace multiple try-with-resources blocks with single compound try-with-resources block

C0urante commented:
Thanks for the review @tombentley, and apologies for the delay. I've addressed your comments and moved the `LogCaptureAppender` class to the clients module where it can be used by both Connect and Streams.

I hope this looks alright given the changes in https://github.com/apache/kafka/pull/11323 that drastically alter the logic touched on here.

We're probably past the point for 3.1 so this isn't particularly urgent anymore but I believe it'd still be useful to have at some point if you can spare the time. No rush, though.

C0urante commented:
CC @vvcephei--this touches on the `LogCaptureAppender` currently used by Streams; you may want to take a look if you have the time.

vvcephei commented on build.gradle:
```diff
@@ -2400,6 +2401,7 @@ project(':connect:runtime') {
     testImplementation project(':core')
     testImplementation project(':metadata')
     testImplementation project(':core').sourceSets.test.output
+    testImplementation project(':streams').sourceSets.test.output
```
Since you've moved the util, do you still need this dependency?

C0urante replied:
Good catch, do not need this!

vvcephei commented on build.gradle:
```diff
@@ -1205,6 +1205,7 @@ project(':clients') {
 
     testImplementation libs.bcpkix
     testImplementation libs.junitJupiter
+    testImplementation libs.log4j
```
I wonder if we can drop this dependency from Streams now. Do you mind checking?

C0urante replied:
Ack, will give it a shot. Haven't tested locally, hoping Jenkins can do that for us.

C0urante replied:
Jenkins has indeed tested for us and the verdict appears to be that this dependency is still required. I that the way that we're pulling in the clients test library prevents transitive dependencies from being included. We might consider changing `testImplementation project(':clients').sourceSets.test.output` to just `testImplementation project(':clients')` but it looks like all non-core projects use that style; probably best to leave that as-is for now and add the log4j dependency back.

vvcephei replied:
Oh, right, I forgot about that. We did that on purpose after something bad got pulled in accidentally. I forget the exact details.

Sorry for the wild goose chase.

ijuma replied:
Hmm, unless clients has a direct dependency on log4j classes, this shouldn't be here. Note that we have the following:

> testRuntimeOnly libs.slf4jlog4j

ijuma replied:
Oh, I see, we moved `LogCaptureAppender` to `clients`. Nevermind.

C0urante replied:
Yeah, the dependency is still necessary. While I was in the neighborhood I noticed that this PR accidentally duplicated the Log4j line, though. Filed #12764 to fix.

vvcephei commented on checkstyle/import-control.xml:
```diff
@@ -531,6 +529,7 @@
       <allow pkg="org.reflections.util"/>
       <allow pkg="javax.crypto"/>
       <allow pkg="org.eclipse.jetty.util" />
+      <allow pkg="org.apache.kafka.streams.processor.internals.testutil" />
```
Interesting. We had an import restriction on pulling in a Streams dependency?

Actually, it looks like the util isn't in this package anymore at all, so this is probably not needed.

C0urante replied:
Yep, good call, thanks for catching this.

I could be wrong but I think the import control logic requires everything to be explicitly allowed; don't think there's an explicit restriction on the Streams package name.

vvcephei reviewed (approved):
Thanks @C0urante !

I just had some small comments. It would be nice to clean those up before merging, but otherwise it LGTM.

tombentley reviewed (approved):
Thanks @C0urante. I guess we might be able to refactor `kafka.utils.LogCaptureAppender` to use the moved `LogCaptureAppender`, but if so we can do that in another PR.

C0urante committed:
Remove unnecessary test dependencies and import allowance, revert some changes from #12434

mimaison commented:
I see 2 committers approved the changes but this was not merged. 
@C0urante can you rebase to fix the conflicts ? Thanks

C0urante commented:
This will create other conflicts with https://github.com/apache/kafka/pull/11780. Would it be possible to prioritize that PR and then, once it's merged, fix the conflicts on this one and merge it?

mimaison commented:
Thanks, that makes sense. I've started reviewing #11780

C0urante commented:
@mimaison I've finally gotten around to rebasing this one, mind taking another look?

mimaison commented:
@tombentley @vvcephei Can you take another look?

C0urante commented:
@mimaison @vvcephei @tombentley I'd like to merge this in order to unblock https://github.com/apache/kafka/pull/12434, which adds a second use case for the `LogCaptureAppender` in the Connect unit tests. Would it be possible to give this another pass sometime next week? Thanks!

C0urante commented:
It's also been proposed in https://github.com/apache/kafka/pull/12434 that we eliminate the Java `LogCaptureAppender` class entirely and replace it with the [Scala-based version](https://github.com/apache/kafka/blob/1cc1e776f703b180f4bd979e8a551805b3bdc94e/core/src/test/scala/unit/kafka/utils/LogCaptureAppender.scala) that's used in core.

Any thoughts?

C0urante commented:
Test failures are unrelated; going to merge this as-is. @vvcephei @tombentley Please let me know if you'd like to follow up on this at any point.
